src/main/java/net/dv8tion/jda/core/requests/Requester.java

```diff
@@ -99,58 +105,67 @@ public JDAImpl getJDA()
-            request = createRequest(route, bodyData);
-        }
+        String url = DISCORD_API_PREFIX + route.getCompiledRoute();
+        builder.url(url);
+
+        String method = apiRequest.getRoute().getMethod().toString();
+        RequestBody body = apiRequest.getData();
+
+        if (body == null && HttpMethod.requiresRequestBody(method))
+            body = EMPTY_BODY;
+
+        builder.method(method, body);
+
+        builder.header("user-agent", USER_AGENT);
+
+        //adding token to all requests to the discord api or cdn pages
+        //we can check for startsWith(DISCORD_API_PREFIX) because the cdn endpoints don't need any kind of authorization
+        if (url.startsWith(DISCORD_API_PREFIX) && api.getToken() != null)
+            builder.header("authorization", api.getToken());
 
         // Apply custom headers like X-Audit-Log-Reason
-        //If customHeaders is null this does nothing
-        request.getHttpRequest().headers(apiRequest.customHeaders);
+        // If customHeaders is null this does nothing
+        if (apiRequest.getHeaders() != null)
+            for (Entry<String, String> header : apiRequest.getHeaders().entrySet())
+                builder.addHeader(header.getKey(), header.getValue());
+
+        okhttp3.Request request = builder.build();
 
         Set<String> rays = new LinkedHashSet<>();
         try
         {
-            HttpResponse<String> response;
+            okhttp3.Response response;
             int attempt = 0;
             do
             {
                 //If the request has been canceled via the Future, don't execute.
                 if (apiRequest.isCanceled())
                     return null;
-                response = request.asString();
-                String cfRay = response.getHeaders().getFirst("CF-RAY");
+                Call call = httpClient.newCall(request);
+                response = call.execute();
+                String cfRay = response.header("CF-RAY");
                 if (cfRay != null)
                     rays.add(cfRay);
 
-                if (response.getStatus() < 500)
+                if (response.code() < 500)
                     break;
 
                 attempt++;
                 LOG.debug(String.format("Requesting %s -> %s returned status %d... retrying (attempt %d)",
-                        request.getHttpRequest().getHttpMethod().name(),
-                        request.getHttpRequest().getUrl(),
-                        response.getStatus(), attempt));
+                        apiRequest.getRoute().getMethod().toString(),
+                        url,
+                        response.code(), attempt));
                 try
                 {
                     Thread.sleep(50 * attempt);
                 }
                 catch (InterruptedException ignored) {}
             }
-            while (attempt < 3 && response.getStatus() >= 500);
+            while (attempt < 3 && response.code() >= 500);
```

Comment:
Turned out this doesn't fully work because request and response bodies are represented as Input-/OutputStreams and as such can only be read once. I can still add a trace log for the other request properties though